### PR TITLE
Check that there are no orphaned .links files in tests directory

### PR DESCRIPTION
**Pull request by frank-emrich**
This is a follow-up to @jamescheney's suggestion in #824 that we should have some way of automatically checking the *.links files in tests/ are somehow executed.

To this end, my script checks that for any .links file in tests/ if it is either
- mentioned in a .tests file
- mentioned in any of the testsuite.config files for database tests
- starts with `#IGNORED_BY_TEST_REGISTRATION_RULE_CHECKING_SCRIPT` to indicate that the file wasn't overlooked, but doesn't have to satisfy the first to conditions (for example, because it is included by another file)

I've also made some tweaks to tools/rule-check, which can now execute any kind of rule checking script (rather than just bash) and also complains if it finds a non-executable file in tools/rules/, which caused #825 and #826.

### CODE REVIEW

File: tools/rules/test_files_registered.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+links_files = set()
+
+#Collect .links files that should appear in a ,tests file
+for links_file in glob.glob(os.path.join(TEST_DIR, "**", "*.links"), recursive=True):
+    with open(links_file) as f:
+        first_line = f.readline().strip()
+        if(first_line != BLACKLIST_STRING):
+            normed = os.path.normpath(links_file)
+            links_files.add(normed)
+            #print("normed : %s" % normed)
```
Comment (SimonJF) [inline tools/rules/test_files_registered.py:18]: Consider deleting

File: tools/rules/test_files_registered.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+            if line == "":
+                test_offset = 0
+            else:
+                test_offset += 1
+
+            #If this is the second line in a test case def, it may be the path of a  .links file
+            if(test_offset == 2 and re.match(".*\.links", line)):
+                normed = os.path.normpath(line)
+                #print("normed: %s" % normed)
```
Comment (SimonJF) [inline tools/rules/test_files_registered.py:39]: Consider deleting

File: tools/rules/test_files_registered.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                if normed in links_files:
+                    #print("found %s" %line)
```
Comment (SimonJF) [inline tools/rules/test_files_registered.py:41]: Consider deleting

File: tools/rules/test_files_registered.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    path_dir = os.path.dirname(testsuite_file)
+
+    with open(testsuite_file, "r") as f:
+        for line in f:
+            path_file = line.strip() + ".links"
+
+            combined_path = os.path.join(path_dir,path_file)
+            normed = normed = os.path.normpath(combined_path)
+            if normed in links_files:
+                    #print("found %s" %line)
```
Comment (SimonJF) [inline tools/rules/test_files_registered.py:57]: Consider deleting

**Comment by SimonJF** [comment]
(The "Consider deleting"s were referring to the commented debug prints, not the entire blocks of code GitHub has chosen to show...)

**Comment by frank-emrich** [comment]
> Have you considered, rather than having a per-file blacklist string, having a JSON config file containing the lists of blacklisted files? 

I've now included the blacklist as a simple list within the script. If we need the list in multiple places, we can still promote it to its own file, but that seemed unnecessary for now.

I've removed the dead print statements

**Comment by SimonJF** [comment]
Looks great to me. Happy for this to go in when you're ready for it to.

**Pull request merged**